File: ios/Runner/MentionTextFieldPlugin.swift
import Flutter
import UIKit

struct MentionUser {
    let id: String
    let name: String
}

struct MentionRange {
    let range: NSRange
    let userId: String
    let userName: String
}

class MentionTextFieldPlugin: NSObject, FlutterPlugin {
    static var channel: FlutterMethodChannel?
    static var views: [Int: MentionTextFieldView] = [:]
    
    static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "com.mybusiness/mention_textfield", binaryMessenger: registrar.messenger())
        let instance = MentionTextFieldPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        
        MentionTextFieldPlugin.channel = channel
        
        let factory = MentionTextFieldFactory(messenger: registrar.messenger())
        registrar.register(factory, withId: "com.mybusiness/mention_textfield")
    }
    
    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "setUsers":
            guard let args = call.arguments as? [String: Any],
                  let viewId = args["viewId"] as? Int,
                  let usersData = args["users"] as? [[String: String]] else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "viewId and users are required", details: nil))
                return
            }
            
            let users = usersData.compactMap { dict -> MentionUser? in
                guard let id = dict["id"], let name = dict["name"] else { return nil }
                return MentionUser(id: id, name: name)
            }
            
            MentionTextFieldPlugin.views[viewId]?.setUsers(users)
            result(true)
            
        case "requestUsers":
            result([])
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    static func notifyTextChanged(viewId: Int, text: String) {
        channel?.invokeMethod("onTextChanged", arguments: ["text": text])
    }
    
    static func notifyTap(viewId: Int) {
        channel?.invokeMethod("onTap", arguments: nil)
    }
}

class MentionTextFieldFactory: NSObject, FlutterPlatformViewFactory {
    private var messenger: FlutterBinaryMessenger
    
    init(messenger: FlutterBinaryMessenger) {
        self.messenger = messenger
        super.init()
    }
    
    func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
        return MentionTextFieldView(frame: frame, viewId: Int(viewId), arguments: args as? [String: Any] ?? [:])
    }
    
    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }
}

class MentionTextFieldView: NSObject, FlutterPlatformView, UITextViewDelegate {
    private var textView: UITextView
    private var viewId: Int
    private var users: [MentionUser] = []
    private var mentions: [MentionRange] = []
    private var mentionTableView: UITableView?
    private var queryStartPos: Int = -1
    private var currentQuery: String = ""
    
    private let mentionColor: UIColor
    private let textColor: UIColor
    
    init(frame: CGRect, viewId: Int, arguments: [String: Any]) {
        self.viewId = viewId
        self.textView = UITextView(frame: frame)
        
        // Parse arguments
        let initialText = arguments["initialText"] as? String ?? ""
        let usersData = arguments["users"] as? [[String: String]] ?? []
        users = usersData.compactMap { dict -> MentionUser? in
            guard let id = dict["id"], let name = dict["name"] else { return nil }
            return MentionUser(id: id, name: name)
        }
        
        textColor = Self.parseColor(arguments["textColor"] as? String ?? "#EAEAEA")
        mentionColor = Self.parseColor(arguments["mentionColor"] as? String ?? "#0095FF")
        let backgroundColor = Self.parseColor(arguments["backgroundColor"] as? String ?? "#1E1E1E")
        let fontSize = CGFloat(arguments["fontSize"] as? Double ?? 14.0)
        
        super.init()
        
        // Configure text view
        textView.delegate = self
        textView.textColor = textColor
        textView.backgroundColor = backgroundColor
        textView.font = UIFont.systemFont(ofSize: fontSize)
        textView.autocorrectionType = .no
        textView.autocapitalizationType = .none
        textView.textContainerInset = UIEdgeInsets(top: 12, left: 12, bottom: 12, right: 12)
        
        // Set initial text
        setTextWithMentions(initialText)
        
        // Register view
        MentionTextFieldPlugin.views[viewId] = self
    }
    
    func view() -> UIView {
        return textView
    }
    
    func setUsers(_ newUsers: [MentionUser]) {
        users = newUsers
    }
    
    // MARK: - UITextViewDelegate
    
    func textViewDidChange(_ textView: UITextView) {
        handleTextChanged()
    }
    
    func textViewDidChangeSelection(_ textView: UITextView) {
        // Handle cursor movement
    }
    
    func textView(_ textView: UITextView, shouldChangeTextIn range: NSRange, replacementText text: String) -> Bool {
        // Check for @ character
        if text == "@" {
            queryStartPos = range.location + 1
            currentQuery = ""
            showMentionDropdown(query: "")
        }
        
        return true
    }
    
    // MARK: - Private Methods
    
    private func handleTextChanged() {
        let text = textView.text ?? ""
        let cursorPos = textView.selectedRange.location
        
        if queryStartPos >= 0 {
            // Currently in mention mode
            if cursorPos > queryStartPos {
                let startIndex = text.index(text.startIndex, offsetBy: queryStartPos)
                let endIndex = text.index(text.startIndex, offsetBy: cursorPos)
                let query = String(text[startIndex..<endIndex])
                
                // Check if query contains space or newline (end of mention)
                if query.contains(" ") || query.contains("\n") {
                    hideMentionDropdown()
                    queryStartPos = -1
                } else {
                    currentQuery = query
                    showMentionDropdown(query: query)
                }
            }
        }
        
        updateMentionFormatting()
        
        // Notify Flutter
        let textWithMentions = getTextWithMentions()
        MentionTextFieldPlugin.notifyTextChanged(viewId: viewId, text: textWithMentions)
    }
    
    private func showMentionDropdown(query: String) {
        let filteredUsers = users.filter { user in
            user.name.lowercased().contains(query.lowercased())
        }
        
        if filteredUsers.isEmpty {
            hideMentionDropdown()
            return
        }
        
        // Create table view if it doesn't exist
        if mentionTableView == nil {
            let tableView = UITableView(frame: CGRect(x: 0, y: 0, width: 200, height: 200))
            tableView.register(UITableViewCell.self, forCellReuseIdentifier: "cell")
            tableView.backgroundColor = UIColor(white: 0.2, alpha: 1.0)
            mentionTableView = tableView
            
            // Add to superview
            if let superview = textView.superview {
                superview.addSubview(tableView)
                
                // Position below text view
                tableView.frame.origin = CGPoint(
                    x: textView.frame.origin.x,
                    y: textView.frame.origin.y + textView.frame.height
                )
            }
        }
        
        // TODO: Implement table view data source and delegate
        // For now, we'll skip the visual dropdown
    }
    
    private func hideMentionDropdown() {
        mentionTableView?.removeFromSuperview()
        mentionTableView = nil
    }
    
    private func insertMention(_ user: MentionUser) {
        let text = textView.text ?? ""
        let cursorPos = textView.selectedRange.location
        
        // Replace from @ to cursor with mention
        let beforeMention = String(text.prefix(queryStartPos - 1))
        let afterMention = String(text.suffix(text.count - cursorPos))
        let mention = "@[\(user.name)](\(user.id))"
        
        let newText = beforeMention + mention + afterMention
        textView.text = newText
        
        // Set cursor position
        let newCursorPos = beforeMention.count + mention.count
        textView.selectedRange = NSRange(location: newCursorPos, length: 0)
        
        // Reset query state
        queryStartPos = -1
        currentQuery = ""
        
        hideMentionDropdown()
        updateMentionFormatting()
    }
    
    private func updateMentionFormatting() {
        guard let text = textView.text else { return }
        
        let attributedString = NSMutableAttributedString(string: text)
        
        // Set default attributes
        attributedString.addAttribute(.foregroundColor, value: textColor, range: NSRange(location: 0, length: text.count))
        attributedString.addAttribute(.font, value: textView.font ?? UIFont.systemFont(ofSize: 14), range: NSRange(location: 0, length: text.count))
        
        // Find and format mentions
        let pattern = "@\\[([^\\]]+)\\]\\(([^)]+)\\)"
        if let regex = try? NSRegularExpression(pattern: pattern, options: []) {
            let matches = regex.matches(in: text, options: [], range: NSRange(location: 0, length: text.count))
            
            for match in matches {
                attributedString.addAttribute(.foregroundColor, value: mentionColor, range: match.range)
                attributedString.addAttribute(.font, value: UIFont.boldSystemFont(ofSize: textView.font?.pointSize ?? 14), range: match.range)
            }
        }
        
        // Preserve cursor position
        let selectedRange = textView.selectedRange
        textView.attributedText = attributedString
        textView.selectedRange = selectedRange
    }
    
    private func getTextWithMentions() -> String {
        return textView.text ?? ""
    }
    
    private func setTextWithMentions(_ text: String) {
        textView.text = text
        updateMentionFormatting()
    }
    
    private static func parseColor(_ hex: String) -> UIColor {
        var hexSanitized = hex.trimmingCharacters(in: .whitespacesAndNewlines)
        hexSanitized = hexSanitized.replacingOccurrences(of: "#", with: "")
        
        var rgb: UInt64 = 0
        Scanner(string: hexSanitized).scanHexInt64(&rgb)
        
        let r = CGFloat((rgb & 0xFF0000) >> 16) / 255.0
        let g = CGFloat((rgb & 0x00FF00) >> 8) / 255.0
        let b = CGFloat(rgb & 0x0000FF) / 255.0
        
        return UIColor(red: r, green: g, blue: b, alpha: 1.0)
    }
}

File: android/app/src/main/kotlin/com/mybusiness/MentionTextFieldPlugin.kt
package com.mybusiness

import android.content.Context
import android.graphics.Color
import android.text.Editable
import android.text.Spannable
import android.text.SpannableStringBuilder
import android.text.TextWatcher
import android.text.style.ForegroundColorSpan
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.PopupWindow
import android.widget.ListView
import android.widget.ArrayAdapter
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory
import io.flutter.embedding.engine.plugins.FlutterPlugin
import java.util.regex.Pattern

data class MentionUser(val id: String, val name: String)

data class MentionRange(val start: Int, val end: Int, val userId: String, val userName: String)

class MentionTextFieldPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private val views = mutableMapOf<Int, MentionTextFieldView>()

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "com.mybusiness/mention_textfield")
        channel.setMethodCallHandler(this)
        
        binding.platformViewRegistry.registerViewFactory(
            "com.mybusiness/mention_textfield",
            MentionTextFieldFactory(this)
        )
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "setUsers" -> {
                val viewId = call.argument<Int>("viewId")
                val usersData = call.argument<List<Map<String, String>>>("users")
                
                if (viewId != null && usersData != null) {
                    val users = usersData.map { MentionUser(it["id"] ?: "", it["name"] ?: "") }
                    views[viewId]?.setUsers(users)
                    result.success(true)
                } else {
                    result.error("INVALID_ARGUMENT", "viewId and users are required", null)
                }
            }
            "requestUsers" -> {
                result.success(emptyList<Map<String, String>>())
            }
            else -> result.notImplemented()
        }
    }

    fun registerView(viewId: Int, view: MentionTextFieldView) {
        views[viewId] = view
    }

    fun unregisterView(viewId: Int) {
        views.remove(viewId)
    }

    fun notifyTextChanged(viewId: Int, text: String) {
        channel.invokeMethod("onTextChanged", mapOf("text" to text))
    }

    fun notifyTap(viewId: Int) {
        channel.invokeMethod("onTap", null)
    }
}

class MentionTextFieldFactory(private val plugin: MentionTextFieldPlugin) : 
    PlatformViewFactory(StandardMessageCodec.INSTANCE) {
    
    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        val creationParams = args as? Map<String, Any> ?: emptyMap()
        return MentionTextFieldView(context, viewId, creationParams, plugin)
    }
}

class MentionTextFieldView(
    context: Context,
    private val viewId: Int,
    creationParams: Map<String, Any>,
    private val plugin: MentionTextFieldPlugin
) : PlatformView {
    
    private val editText: EditText = EditText(context)
    private var users = listOf<MentionUser>()
    private val mentions = mutableListOf<MentionRange>()
    private var mentionPopup: PopupWindow? = null
    private var queryStartPos = -1
    private var currentQuery = ""
    
    private val mentionColor: Int
    private val textColor: Int
    
    init {
        plugin.registerView(viewId, this)
        
        // Parse creation params
        val initialText = creationParams["initialText"] as? String ?: ""
        val usersData = creationParams["users"] as? List<Map<String, String>> ?: emptyList()
        users = usersData.map { MentionUser(it["id"] ?: "", it["name"] ?: "") }
        
        textColor = parseColor(creationParams["textColor"] as? String ?: "#EAEAEA")
        mentionColor = parseColor(creationParams["mentionColor"] as? String ?: "#0095FF")
        val backgroundColor = parseColor(creationParams["backgroundColor"] as? String ?: "#1E1E1E")
        val fontSize = (creationParams["fontSize"] as? Double ?: 14.0).toFloat()
        val hintText = creationParams["hintText"] as? String ?: "Digite o texto..."
        
        // Configure EditText
        editText.apply {
            setText(initialText)
            setTextColor(textColor)
            setBackgroundColor(backgroundColor)
            textSize = fontSize
            hint = hintText
            setHintTextColor(Color.parseColor("#888888"))
            setPadding(12, 12, 12, 12)
            
            layoutParams = ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT
            )
        }
        
        // Add text watcher
        editText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                handleTextChanged(s?.toString() ?: "")
            }
            
            override fun afterTextChanged(s: Editable?) {
                updateMentionFormatting()
            }
        })
        
        // Set initial text with mentions
        setTextWithMentions(initialText)
        
        editText.setOnClickListener {
            plugin.notifyTap(viewId)
        }
    }
    
    override fun getView(): View = editText
    
    override fun dispose() {
        plugin.unregisterView(viewId)
        mentionPopup?.dismiss()
    }
    
    fun setUsers(newUsers: List<MentionUser>) {
        users = newUsers
    }
    
    private fun handleTextChanged(text: String) {
        // Check for @ character to start mention
        val cursorPos = editText.selectionStart
        
        if (queryStartPos >= 0) {
            // Currently in mention mode
            if (cursorPos > queryStartPos) {
                val query = text.substring(queryStartPos, cursorPos)
                
                // Check if query contains space or newline (end of mention)
                if (query.contains(' ') || query.contains('\n')) {
                    hideMentionDropdown()
                    queryStartPos = -1
                } else {
                    currentQuery = query
                    showMentionDropdown(query)
                }
            }
        } else {
            // Check if @ was just typed
            if (cursorPos > 0 && text.getOrNull(cursorPos - 1) == '@') {
                queryStartPos = cursorPos
                currentQuery = ""
                showMentionDropdown("")
            }
        }
        
        // Notify Flutter
        val textWithMentions = getTextWithMentions()
        plugin.notifyTextChanged(viewId, textWithMentions)
    }
    
    private fun showMentionDropdown(query: String) {
        val filteredUsers = users.filter { 
            it.name.contains(query, ignoreCase = true) 
        }
        
        if (filteredUsers.isEmpty()) {
            hideMentionDropdown()
            return
        }
        
        // Create popup if it doesn't exist
        if (mentionPopup == null) {
            val listView = ListView(editText.context)
            mentionPopup = PopupWindow(
                listView,
                ViewGroup.LayoutParams.WRAP_CONTENT,
                ViewGroup.LayoutParams.WRAP_CONTENT,
                true
            )
        }
        
        val listView = mentionPopup?.contentView as? ListView
        listView?.adapter = ArrayAdapter(
            editText.context,
            android.R.layout.simple_list_item_1,
            filteredUsers.map { it.name }
        )
        
        listView?.setOnItemClickListener { _, _, position, _ ->
            insertMention(filteredUsers[position])
        }
        
        // Show popup below cursor
        mentionPopup?.showAsDropDown(editText)
    }
    
    private fun hideMentionDropdown() {
        mentionPopup?.dismiss()
    }
    
    private fun insertMention(user: MentionUser) {
        val text = editText.text.toString()
        val cursorPos = editText.selectionStart
        
        // Replace from @ to cursor with mention
        val beforeMention = text.substring(0, queryStartPos - 1)
        val afterMention = text.substring(cursorPos)
        val mention = "@[${user.name}](${user.id})"
        
        val newText = beforeMention + mention + afterMention
        editText.setText(newText)
        editText.setSelection(beforeMention.length + mention.length)
        
        // Reset query state
        queryStartPos = -1
        currentQuery = ""
        
        hideMentionDropdown()
    }
    
    private fun updateMentionFormatting() {
        val text = editText.text
        if (text !is Spannable) return
        
        // Remove existing spans
        val spans = text.getSpans(0, text.length, ForegroundColorSpan::class.java)
        spans.forEach { text.removeSpan(it) }
        
        // Find and format mentions
        val pattern = Pattern.compile("@\\[([^\\]]+)\\]\\(([^)]+)\\)")
        val matcher = pattern.matcher(text)
        
        while (matcher.find()) {
            val start = matcher.start()
            val end = matcher.end()
            text.setSpan(
                ForegroundColorSpan(mentionColor),
                start,
                end,
                Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
            )
        }
    }
    
    private fun getTextWithMentions(): String {
        return editText.text.toString()
    }
    
    private fun setTextWithMentions(text: String) {
        editText.setText(text)
        updateMentionFormatting()
    }
    
    private fun parseColor(hex: String): Int {
        return try {
            Color.parseColor(hex)
        } catch (e: Exception) {
            Color.parseColor("#EAEAEA")
        }
    }
}

